### PR TITLE
Evaluation Routine Implementation

### PR DESCRIPTION
**Pull request by nquetschlich**
This PR...
- adds the supermarq features (already used in the MQT Predictor) to this repository as well
- adds an `evaluation` module including visualization generation routines and functions to calculate different characteristics
- adds `nb-clean` to the pre-commit hooks

### CODE REVIEW

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:10]: seems like a leftover to me

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:35]: How about making that return type a `TypedDict` so that the individual evaluation metrics also get a meaning assigned to them?

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:32]: I am not a big fan of dumping these results as part of the package since it needlessly increases the package size by 6.5MB. 
How long does the evaluation of these files take in general?
Is it feasible to just return `res` in this function?

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:63]: Isn't that function already defined somewhere?
I vaguely remember seeing something like this. But maybe it was in the Predictor.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:62]: That error message could be a little more descriptive. As it is, it contains almost no information about the underlying problem.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:121]: Is that also "just" taken from the predictor?
If so, then consider moving it to an appropriate file here as well. IIRC, it's something like `reward` or so over there.
It certainly makes sense to move some common functions from the predictor over here.
Consequently, they should be removed from the Predictor in a subsequent update to avoid code duplication.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation_visualization.ipynb:28]: Already stated it above, but would it be too time consuming to perform the calculation of the evaluation data every time?
That would avoid having to store these pickle files.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation_visualization.ipynb:42]: I believe this code could be simplified by introducing the TypedDict as proposed above.
Because `eval_data` then is a simple list of these TypedDicts.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation_visualization.ipynb:116]: I'd say this "filename" to "counter dictionary" transformation could be packaged as a function in the evaluation module.
I'd put slightly more focus on making the method extensible with future architectures and gate-sets. Maybe TypedDicts or a slightly more sophisticated matching could help here as well.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation_visualization.ipynb:186]: I just put this comment here because it is the first place where this suggestion fits, but what's coming next applies way more broadly to you repositories:

I am not a big fan of storing many pre-generated PDFs as part of our repositories as they quickly amount to a lot of files and just "lie around" in folders without much context.
Even more so, I am not a huge fan of Jupyter notebooks as they are really hard to read when not rendered properly by something like GitHub or your local IDE.

So the first thing I would advocate for, is moving as many utility functions as possible from the Jupyter notebook to a simple Python module. In this case here, that could be the `evaluation` module. Or it could be called `visualization` or something similar.
It would also be great if these visualisation methods could be somewhat covered with unit tests. That would probably require some kind of mocking/monkeypatching, but should be possible.
IMHO, the Jupyter notebooks should merely consist of some calls to library functions plus some added text for documentation purposes.

Now, I am a big fan of how we use Jupyter Notebooks in projects like QCEC and QMAP, where they are part of the documentation. See, e.g., https://github.com/cda-tum/qcec/blob/main/docs/source/Quickstart.ipynb, which gets rendered as https://qcec.readthedocs.io/en/latest/Quickstart.html in the documentation.
What happens here, is that the RtD documentation builder just executes the (clean) Jupyter notebooks and builds a documentation page out of that; including all the output.

All the evaluations we perform here (but also in the Predictor or the Problem Solver) would very nicely fit into their own documentation pages (assuming the generation of the respective images does not take hours).
This would eliminate the need for storing the individual PDFs in the repository as they would just be generated on-the-fly. In addition, documentation pages provide quite a lot more context than the individual PDF files lying around.

I guess what I am trying to say is: We should really get some RtD up and running cause it could significantly improve the way we can visualize data and results.

File: tests/test_bench.py
Comment (burgholzer) [inline tests/test_bench.py:782]: I'd say most of these assertions become redundant when the evaluation returns a TypedDict.
These tests could be a little more comprehensive I guess.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Thanks for this PR!
As (almost) always, I have a couple of thoughts on this or it triggered some thoughts.
These thoughts are best summarised in the comments below, but I think the gist of it is:

I believe there is merit in having these kind of evaluations and visualisations as part of the repository. However, I feel that the current way is not really sustainable when thinking about the future and needs to be automated to be practical. Such an automation could happen in the form of proper RtD documentation and some generalisations of the functions used to generate the visualisations.

I do think that these documentation changes deserve their own PR though and the PR here can be merged as an intermediate step, after all the other comments have been addressed.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py:32]: Unfortunately, the generation takes around 11h. Therefore returning `res` might not be a viable approach.

Comment (burgholzer) [reply src/mqt/bench/evaluation/evaluation.py:32]: Ok. Alternative suggestion: Can we dynamically download the `evaluation_data.pkl` from release assets just like we do with the `.zip`?

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Removed that fragment.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Good point, implemented such a `TypedDict`.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Yes, removed it according to the MQT Predictor and will delete it there in the next days.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Yes, removed it according to the MQT Predictor and will delete it there in the next days.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Improved on that.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation_visualization.ipynb:42]: Improved that.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation_visualization.ipynb:186]: I agree with you. The documentation improvement and shift to Read-the-Docs is on my todo-list for the next weeks.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [reply src/mqt/bench/evaluation/evaluation.py]: Sounds good. Keep in mind that a new MQT Bench release is necessary for the functions to be available in the Predictor and that the lower version cap of MQT in the Predictor needs to be increased accordingly.

File: src/mqt/bench/utils.py
Comment (burgholzer) [inline src/mqt/bench/utils.py:348]: Why are you annotating the return type with `Any` here? Shouldn't that just be `int`?

File: src/mqt/bench/utils.py
Comment (burgholzer) [inline src/mqt/bench/utils.py:362]: Also a nice opportunity to introduce a TypedDict, or probably even better: a dataclass.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:38]: Now that I think of it, a `dataclass` is probably more suitable here since the dictionary just creates an unnecessary overhead compared to the `tuple`.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Good points. I created an issue (https://github.com/cda-tum/MQTPredictor/issues/74) in the Predictor to be reminded of that.

File: src/mqt/bench/evaluation/evaluation_visualization.ipynb
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation_visualization.ipynb:125]: I added two counting methods in the `evaluation.py` file to remove most of the calculations (but simple divisions ) from the jupyter notebook

File: src/mqt/bench/utils.py
Comment (nquetschlich) [reply src/mqt/bench/utils.py]: Fixed that.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:59]: No need to cast here. Just do
```suggestion
    return EvaluationResult(
        filename=filename,
        num_qubits=int(str(filename).split("_")[-1].split(".")[0]),
        depth=qc.depth(),
        etc.
    )
```
And at that point, you can use a data class as well, which would look exactly the same here.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:67]: These functions aren't covered by tests. That should be fairly easy to change.

File: src/mqt/bench/utils.py
Comment (burgholzer) [inline src/mqt/bench/utils.py:348]: Coming back to this again: Why is the `qargs` argument typed as `list[Any]` instead of a proper type?

File: tests/test_bench.py
Comment (burgholzer) [inline tests/test_bench.py:777]: the type isn't only a dict. it has to be the typed dict you created.

**Review by burgholzer — CHANGES_REQUESTED** [review]
Just some further comments here because I had some time to go through this.

File: src/mqt/bench/utils.py
Comment (burgholzer) [inline src/mqt/bench/utils.py:356]: While this error message is better than the message before, it still contains absolutely no information about the local qubit index that was the cause of the error.
```suggestion
    error_msg = f"Global qubit index for local qubit {index} index not found."
```

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Implemented it now as a dataclass

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Adjusted the code accordingly.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py:68]: Added respective tests.

File: tests/test_bench.py
Comment (nquetschlich) [reply tests/test_bench.py]: Changed that.

File: src/mqt/bench/utils.py
Comment (nquetschlich) [reply src/mqt/bench/utils.py]: tbh, because I have overlooked that. Added it by now.

File: src/mqt/bench/utils.py
Comment (nquetschlich) [reply src/mqt/bench/utils.py:377]: Implemented it.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:36]: Wrong type ;o)
We really need mypy here as well.

Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Fixed that...

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: I removed the file from the repo as discussed.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py]: I guess this is still some leftover from debugging?
This should probably use logging instead of print. Fine for this PR, but maybe to keep in mind.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:42]: Can just be inlined below.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:25]:
```suggestion
    with (resources.files("mqt.bench") / "evaluation" / "evaluation_data.pkl").open("wb") as f:
```
Should also work and be more compact.

File: src/mqt/bench/utils.py
Comment (burgholzer) [inline src/mqt/bench/utils.py:397]: Needs a proper type ;o)
Shouldn't be too hard to infer.

File: tests/test_bench.py
Comment (burgholzer) [inline tests/test_bench.py:768]: Seems like this is not needed anymore

**Review by burgholzer — CHANGES_REQUESTED** [review]
A last round of review before this can be merged. Just minor details.

File: src/mqt/bench/evaluation/evaluation.py
Comment (burgholzer) [inline src/mqt/bench/evaluation/evaluation.py:61]: What is this lonely call doing here? 🙃

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: True, removed that.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: Done.

File: tests/test_bench.py
Comment (nquetschlich) [reply tests/test_bench.py]: Removed it.

File: src/mqt/bench/evaluation/evaluation.py
Comment (nquetschlich) [reply src/mqt/bench/evaluation/evaluation.py]: That happened accidentally.

File: src/mqt/bench/utils.py
Comment (nquetschlich) [reply src/mqt/bench/utils.py]: Yes, was quite simple. Additionally, I improved the `connectivity` initialization routine below.Yes, was quite simple. Additionally, I improved the `connectivity` initialization routine below.

**Review by burgholzer — APPROVED** [review]
LGTM 👍🏼

**Pull request merged**
